### PR TITLE
Add concept exercise objects

### PR DESCRIPTION
**Pull request by junedev**
Closes https://github.com/exercism/javascript/issues/993

This one was harder than the others ones. I feel it could still need some more work (left some notes in design.md) but I want to get it out of the door. Hope it is ok for now.

### CODE REVIEW

File: exercises/concept/high-score-board/.meta/exemplar.js
Comment (junedev) [inline exercises/concept/high-score-board/.meta/exemplar.js:6]: @SleeplessByte I am not up to date with this JSDoc ts-check business. :point_up_2:  This is what I found in the JSDoc documentation how to describe the object. Is that ok or should there be something else (more typescripty) here?

Comment (SleeplessByte) [reply exercises/concept/high-score-board/.meta/exemplar.js:6]: @junedev you want to represent a shape that has keys of string and values of number?

```typescript
@returns {Record<string, number>}
```

which is essentially the same as:

```typescript
@returns {{ [key: string]: number }}
```

Comment (junedev) [reply exercises/concept/high-score-board/.meta/exemplar.js]: > you want to represent a shape that has keys of string and values of number?

Yes. Inserted your suggestion. Thanks!

File: concepts/objects/about.md
Comment (SaschaMann) [inline concepts/objects/about.md:7]:
```suggestion
In other languages they are referred to as maps or dictionaries and often times all values need to have the same data type.
```

File: concepts/objects/about.md
Comment (SaschaMann) [inline concepts/objects/about.md:14]: Does the student know what `(literal)` refers to at this point in the track?

If not, perhaps adding a brief footnote or explanation would prevent some web searching.

File: concepts/objects/about.md
Comment (SaschaMann) [reply concepts/objects/about.md:7]: My initial guess would have been that they're more similar to structs or data classes in other languages, which can have varying value types. Perhaps you could briefly mention how those two relate?

File: concepts/objects/about.md
Comment (SaschaMann) [inline concepts/objects/about.md:55]:
```suggestion
JavaScript has a short-hand notation that allows just writing `name` instead.
```

`allow` is followed by a gerund, I think.

File: concepts/objects/about.md
Comment (SaschaMann) [inline concepts/objects/about.md:36]: It may be a good idea to mention why one would want to add a trailing comma (better diffs etc).

File: concepts/objects/about.md
Comment (SaschaMann) [inline concepts/objects/about.md:114]:
```suggestion
See [Null and Undefined][concept-null-undefined] for more details on this.
```

File: concepts/objects/about.md
Comment (SaschaMann) [inline concepts/objects/about.md:207]:
```suggestion
The `for…in` loop has a pitfall in case the object has [inherited keys][concept-inheritance].
```

File: concepts/objects/about.md
Comment (SaschaMann) [inline concepts/objects/about.md:208]:
```suggestion
Those will also be visited by the `for…in` loop.
```

File: concepts/objects/about.md
Comment (SaschaMann) [inline concepts/objects/about.md:187]: Are objects in JS iterables? Would `for…of` work too?

File: concepts/objects/introduction.md
Comment (SaschaMann) [inline concepts/objects/introduction.md:1]: Imo the headers from the about should be in this doc, too. Makes it easier to navigate since it's quite long.

File: concepts/objects/about.md
Comment (SaschaMann) [reply concepts/objects/about.md:207]: Not sure which one is the common name, but I think consistency might be nicer. Same applies to the introduction.

File: exercises/concept/high-score-board/.docs/hints.md
Comment (SaschaMann) [inline exercises/concept/high-score-board/.docs/hints.md:5]:
```suggestion
- Create a new object with curly brackets.
```

consistency with the other hints

File: exercises/concept/high-score-board/.docs/instructions.md
Comment (SaschaMann) [inline exercises/concept/high-score-board/.docs/instructions.md:37]: Very minor/subjective:

This somewhat sounds like the task would be about implementing logic that detects old scores and removes them. It's not misleading since the rest of the task makes it clear what it's about but perhaps you could avoid this by changing it to something like "If a player was found to be cheating, they are manually removed from the high score board."

File: exercises/concept/high-score-board/.meta/config.json
Comment (SaschaMann) [inline exercises/concept/high-score-board/.meta/config.json:10]:
```suggestion
  "forked_from": ["julia/high-score"]
```

File: exercises/concept/high-score-board/.meta/design.md
Comment (SaschaMann) [inline exercises/concept/high-score-board/.meta/design.md:6]: To me their importance didn't come across from the exercise docs. Perhaps ask others for input on this, though, before adding more info that might just end up being redundant.

File: exercises/concept/high-score-board/.meta/design.md
Comment (SaschaMann) [inline exercises/concept/high-score-board/.meta/design.md:9]:
```suggestion
- What keys are allowed
- What values are allowed (e.g., numbers, strings etc or other objects, arrays or even functions)
```

File: exercises/concept/high-score-board/.meta/design.md
Comment (SaschaMann) [inline exercises/concept/high-score-board/.meta/design.md:15]:
```suggestion
- How to iterate through the keys with `for...in`
```

File: config.json
Comment (SaschaMann) [inline config.json:176]: `strings` and `numbers` might be pre-reqs too but I don't know enough about the JS progression. Just pointing it out for someone else's review.

File: exercises/concept/high-score-board/.meta/design.md
Comment (SaschaMann) [inline exercises/concept/high-score-board/.meta/design.md:51]:
```suggestion
   - TODO/FIXME
```

Again, not sure what the conventions are in the JS repo, but `???` isn't easily findable via grepping, I'm guessing.

File: exercises/concept/high-score-board/.meta/design.md
Comment (SaschaMann) [inline exercises/concept/high-score-board/.meta/design.md:55]:
```suggestion
   - `essential`: Make sure `delete` was used and not set to undefined or null.
```

File: exercises/concept/high-score-board/.meta/design.md
Comment (SaschaMann) [inline exercises/concept/high-score-board/.meta/design.md:75]: `forked_from` points to the Julia implementation, and `source` is empty. Perhaps add Elixir to either of those?

I'm not the best person to ask about this tho, @iHiD will likely have a different stance on it.

File: exercises/concept/high-score-board/LICENSE
Comment (SaschaMann) [inline exercises/concept/high-score-board/LICENSE:3]:
```suggestion
Copyright (c) 2021 Exercism
```

**Review by SaschaMann — COMMENTED** [review]
Looks great overall, I left some minor comments (from the PoV of someone who barely knows JS). I did not look at the tests/exemplar files.

File: concepts/objects/about.md
Comment (iHiD) [reply concepts/objects/about.md:7]: Also, on this line "often times" isn't familiar sounding English to me. Just "often" without the "times" is clearer 🙂

Comment (SaschaMann) [reply concepts/objects/about.md:7]: I think it's a US phrase mostly

File: exercises/concept/high-score-board/LICENSE
Comment (SleeplessByte) [reply exercises/concept/high-score-board/LICENSE:3]: This needs to be fixed in the ROOT, and then it propagates via a sync. Probably a separate PR :)

File: concepts/objects/about.md
Comment (SleeplessByte) [inline concepts/objects/about.md:57]: Might not be the place to mention this, but the short-hand notation is **not** equivalent to the key-value notation. The difference is small, but there:

```javascript
const appleTree = {
  kind: 'apple',
  getParent: function() {
    return super.toString();
  }
};
// Uncaught SyntaxError: 'super' keyword unexpected here

const pearTree = {
  kind: 'pear',
  getParent() {
    return super.toString();
  }
};
// this works

// You can then declare things on its prototype
Object.setPrototypeOf(pearTree, { toString() { return "BigTree" } })
pearTree.getParent()
// => "BigTree"
```

It's a weird edgy case, but it matters as soon as you're going to type to _copy_ that method you've declared or if you're trying to write mixins.

Comment (SleeplessByte) [reply concepts/objects/about.md:57]: There's another _difference_ between these:

```javascript
const obj = {
  calcSum: function calc(a, b) {
    return a + b;
  },
};

obj.calcSum.name
// => "calc"

const obj2 = {
  calcSum: function (a, b) {
    return a + b;
  },
};

obj2.calcSum.name
// => "calcSum"

const obj3 = {
  calcSum(a, b) {
    return a + b;
  },
};

obj3.calcSum.name
// => "calcSum"
```

It's pretty insignificant, but 🤷🏽

File: concepts/objects/about.md
Comment (SleeplessByte) [inline concepts/objects/about.md:89]: Might be worth showing the result

File: concepts/objects/about.md
Comment (SleeplessByte) [inline concepts/objects/about.md:93]: We should definitely NOT mention this, but there are more than two ways.

```javascript
const obj = { greeting: 'hello world' }

const descriptor = Object.getOwnPropertyDescriptor(obj , 'greeting')
// => {value: "hello world", writable: true, enumerable: true, configurable: true}

descriptor.value
// => "hello world"
```

And naturally by way of destructuring.

There's even `Object.values`:

```javascript
const obj = { foo: 'bar', greeting: 'hello world' }

const values = Object.values(obj)
const index = Object.getOwnPropertyNames(obj).indexOf('greeting')

values[index]
// => "hello world"
```

By placing these in this comment on this PR, someone, someday might find them and be sad or happy.

File: concepts/objects/about.md
Comment (SleeplessByte) [inline concepts/objects/about.md:131]: Perhaps worth mentioning that you _could_ mix and match?

```javascript
obj.address['city']
```

File: concepts/objects/about.md
Comment (SleeplessByte) [inline concepts/objects/about.md:149]: If you add the property using `Object.create` or like this:

```javascript
const obj = {}
Object.defineProperty(obj, 'greeting', { value: 'hello world', configurable: true })
// => {greeting: "hello world"}
```

You can actually then do:

```javascript
Object.defineProperty(obj, 'greeting', { value: 'goodbye world' })
// => {greeting: "goodbye world"}
```

`defineProperty` etc is probably something we want somewhere but not here, but I don't know where so that's why I put it here so I hope it gives you inspiration.

File: concepts/objects/about.md
Comment (SleeplessByte) [inline concepts/objects/about.md:165]: Very good advice!

Might be worth showing `hasOwnProperty` to validate something is gone, given that you show it below?

File: concepts/objects/about.md
Comment (SleeplessByte) [inline concepts/objects/about.md:183]: You might want to mention `getOwnPropertyNames` which are all the properties `hasOwnProperty` is true for.

File: concepts/objects/about.md
Comment (SleeplessByte) [reply concepts/objects/about.md:187]: `for...in` and `for...of` are not equivalent, and the concept of enumerable is not the same as iterable, in JavaScript (@junedev all this might be worth mentioning, somewhere).

JavaScript objects have _enumerable properties_. They are enumerable by default, and can be made non enumerable using `Object.create` or `Object.defineProperty` setting `{ enumerable: false }`.

They are _only_ iterable if there is `Symbol.iterator` defined (as a method):

```javascript
const iterable1 = {};

iterable1[Symbol.iterator] = function* () {
  yield 1;
  yield 2;
  yield 3;
};

console.log([...iterable1]);
// [1, 2, 3]
```

[Array's prototype has it defined](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/@@iterator), but Object's prototype does not. _Iterability_ is **only** used by `for ... of`. The async variant is `Symbol.asyncIterator` and is used by **for-await-of**.

File: concepts/objects/about.md
Comment (SleeplessByte) [inline concepts/objects/about.md:209]: Perhaps it should be mentioned that you almost _always_ want this.

**Review by SleeplessByte — COMMENTED** [review]
I've only reviewed about.md, because I don't want to overload ya here, but see if it aids you in any way.

File: concepts/objects/about.md
Comment (junedev) [reply concepts/objects/about.md:57]: Re the first issue: 
Interesting, I didn't know and it was not mentioned [on MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Method_definitions#description). Do you know why there is a difference or can you provide some link where I could read up on this? Tbh I would not hold off merging the PR until I got in. But I can put it into the design document as improvement idea. It might be a good fit once we have the special "sections" for about.md where you can mark something as advanced content.

Re the second issue:
True, but this is more about how function names are determined than about the shorthand notation imo. It could go into the functions concept but even there we have a lot more basic ground to cover first.

Its not worth holding off this PR for any of those. I would put both things into the design.md improvement ideas section of the respective exercise. They might be a good fit once we have the special "sections" for about.md where you can mark something as advanced content.

File: concepts/objects/about.md
Comment (junedev) [reply concepts/objects/about.md:93]: "There are some more ways ... " probably applies to every other thing I cover in all the JavaScript concepts. :wink:

File: concepts/objects/about.md
Comment (junedev) [reply concepts/objects/about.md:149]: Totally agree it should go in somewhere, that is why I wrote "Helper methods like `Object.assign` etc." should be covered in the about.md in the future in the design document. I rather wanted to move on to other concepts at some point. So in a next iteration on this I would see these helpers being explained in the `objects` about.md and then further down the road they could even be moved into some "objects advanced" or whatever exercise if we find more related things we maybe want to cover with real exercises and not just mention in about.md.

File: config.json
Comment (junedev) [reply config.json:176]: We discussed that we don't add the primitives types as prerequisites further down the concept tree in the JS track if they are not a core part of the current concept.

File: concepts/objects/about.md
Comment (junedev) [reply concepts/objects/about.md:7]: @SaschaMann you are totally right with that first comment, in my head it was clear that we will cover that struct/class aspect of objects elsewhere but I forgot to also make this clear for the reader. Rewrote that section. Removed the "times" as well.

File: concepts/objects/about.md
Comment (junedev) [reply concepts/objects/about.md:14]: Good point. Jeremy recommended to use links in these cases where we use terms that someone that already knows programming should know but in reality might not know.

File: concepts/objects/about.md
Comment (junedev) [reply concepts/objects/about.md:36]: Provided a link where this is spelled out. I think that is sufficient.

File: exercises/concept/high-score-board/.meta/design.md
Comment (junedev) [reply exercises/concept/high-score-board/.meta/design.md:51]: true, but I had in idea what to write there now so this became obsolete

File: exercises/concept/high-score-board/.meta/design.md
Comment (junedev) [reply exercises/concept/high-score-board/.meta/design.md:75]: I just saw in the "stories" docs that Elixir is actually the reference implementation. Used that one as forked_from and in the notes now.

File: concepts/objects/about.md
Comment (junedev) [reply concepts/objects/about.md:10]: I meant is as "but also entities like arrays" not as "but also types like arrays". I think it is ok as is. The "object basics" concept is not the right point for the "well actually functions are objects" point.

File: concepts/objects/about.md
Comment (junedev) [reply concepts/objects/about.md:165]: I think just showing it below is enough here.

File: concepts/objects/about.md
Comment (junedev) [reply concepts/objects/about.md:183]: Same category as "defineProperty", I added it the improvement ideas list.

File: concepts/objects/about.md
Comment (junedev) [reply concepts/objects/about.md:187]: @SleeplessByte I would not go into the details of this in this concept. Left a note (and the link to this comment) on our concept tracking issue. https://github.com/exercism/javascript/issues/951#issuecomment-864452573

File: concepts/objects/about.md
Comment (junedev) [reply concepts/objects/about.md:209]: Good point, I changed the wording accordingly.

File: exercises/concept/high-score-board/.docs/instructions.md
Comment (junedev) [reply exercises/concept/high-score-board/.docs/instructions.md:37]: I realized the same thing at some point but forgot to fix it then. Did it now.

File: exercises/concept/high-score-board/.meta/config.json
Comment (junedev) [reply exercises/concept/high-score-board/.meta/config.json:10]: see https://github.com/exercism/javascript/pull/1160#discussion_r654829051

File: exercises/concept/high-score-board/.meta/design.md
Comment (junedev) [reply exercises/concept/high-score-board/.meta/design.md:6]: I removed this line. I was hinting at the common saying "everything is an object" in JavaScript but I found out that this is partly a misconception when I was looking into it in more detail. :see_no_evil:

File: concepts/objects/introduction.md
Comment (junedev) [reply concepts/objects/introduction.md:1]: You are totally right. Fixed.

**Comment by junedev** [comment]
> I've only reviewed about.md, because I don't want to overload ya here, but see if it aids you in any way.
@SleeplessByte No need to spare me. :wink: I addressed the existing comments by you and Sascha. Do you want to also review the rest?

File: concepts/objects/about.md
Comment (SaschaMann) [inline concepts/objects/about.md:15]:
```suggestion
In the following we will focus on objects as collections or maps. Other use cases of objects will be covered in other concepts, see e.g. [Classes][concept-classes].
```

File: concepts/objects/about.md
Comment (SaschaMann) [reply concepts/objects/about.md:198]: Depending on what comes first in the track, this may be worth mentioning, i.e. if `for..of` is explained in a previous exercise. That way, the student knows to watch out that `for..of` won't work here.

Thanks for the explanation :)

File: concepts/objects/about.md
Comment (SaschaMann) [inline concepts/objects/about.md:7]:
```suggestion
As such, they can be used as what is often referred to as maps or dictionaries in other languages.
```

structs in other languages are also sometimes used purely as a collection. It's rare but imo making this statement slightly weaker clears up that it doesn't _always_ translate to a map or dictionary even if in many languages it does.

The difference to classes with methods in the context of this exercise is clearer to me now, though, so I think it'd be fine as it is too.

**Review by SaschaMann — APPROVED** [review]

**Comment by SaschaMann** [comment]
(turns out I still have write access to this repo, so this is mergeable now, but it's probably best to wait for DJ's proper approval anyway. I meant to approve in the sense that my comments have all been addressed.)

File: concepts/objects/about.md
Comment (junedev) [reply concepts/objects/about.md:15]: @SaschaMann Are you sure about this? I read [here](https://exercism.lol/docs/building/markdown/style-guide) that we use American English and from what read there you always put in the comma. 🤔

File: concepts/objects/about.md
Comment (junedev) [reply concepts/objects/about.md]: fixed in all three places

File: concepts/objects/about.md
Comment (junedev) [reply concepts/objects/about.md:198]: `for...of` comes later in our current unlock flow but we treat about.md as independent from that flow. So your point is totally valid. I added a small part on this. I used a link to the MDN docs for now. Once we have our own concept on enumerable and iterable we can link that instead.

File: exercises/concept/high-score-board/.meta/exemplar.js
Comment (SleeplessByte) [inline exercises/concept/high-score-board/.meta/exemplar.js:75]:
```suggestion
  return params.normalizeFunction(params.score);
```

File: exercises/concept/high-score-board/.meta/exemplar.js
Comment (SleeplessByte) [inline exercises/concept/high-score-board/.meta/exemplar.js:71]:
```suggestion
 * @param {function} params.normalizeFunction
```

File: exercises/concept/high-score-board/high-score-board.js
Comment (SleeplessByte) [inline exercises/concept/high-score-board/high-score-board.js:62]:
```suggestion
 * @param {function} params.normalizeFunction
```

**Review by SleeplessByte — APPROVED** [review]
Looks good. Few nitpicks.

File: concepts/objects/about.md
Comment (SaschaMann) [reply concepts/objects/about.md:15]: I have never seen a comma after `e.g.`, hence my suggestion, but apparently you're right and American English puts a comma there indeed. The doc you linked to gets that wrong too funnily enough.

The more you know... :D

File: exercises/concept/high-score-board/.docs/hints.md
Comment (junedev) [reply exercises/concept/high-score-board/.docs/hints.md:18]: I'm sorry but I don't understand the point about not using "you" in the hints. This is the "official" example from https://exercism.lol/docs/building/tracks/concept-exercises. I used that same style.
 ```markdown
 ## General

- You need to define a [constant][constant] which should contain the [integer][integers] value specified in the recipe.

## 1. Calculate the remaining oven time in minutes

- You need to define a [method][methods] with a single parameter for the actual time so far.
```
Are the docs wrong here?

File: exercises/concept/high-score-board/.docs/instructions.md
Comment (junedev) [reply exercises/concept/high-score-board/.docs/instructions.md:12]: rephrased it completely

File: exercises/concept/high-score-board/.docs/instructions.md
Comment (junedev) [reply exercises/concept/high-score-board/.docs/instructions.md:99]: :+1: Fixed.

File: exercises/concept/high-score-board/.meta/exemplar.js
Comment (junedev) [reply exercises/concept/high-score-board/.meta/exemplar.js:69]: done

**Comment by junedev** [comment]
@SleeplessByte I fixed all the things besides the "you" issue, see [comment](https://github.com/exercism/javascript/pull/1160/files#r656490705).

File: exercises/concept/high-score-board/.docs/hints.md
Comment (SleeplessByte) [reply exercises/concept/high-score-board/.docs/hints.md:18]: The only reason I remove you when refering to a previous exercise or task is because they might not have done that themselves.

It's not a hard requirement, but a nice thing to do.

"You should know how to do this because you did it previously" becomes "it can be done like it has been done previously".

If the student asked for help on the previous implementation, they might go back to the person they've asked help from, without another negative pointer.

Not a blocker, feel free to ignore.

Comment (junedev) [reply exercises/concept/high-score-board/.docs/hints.md:18]: Ah ok, I misunderstood your comments/suggestions as "remove all the 'you'".

Nevertheless I would like to choose "ignore" here because of the "prefer active over passive voice" rule that was stated somewhere (can't find the reference anymore, wasn't moved to /docs on v3 website yet). That said, I still don't have merge rights so I need to you click the button. :see_no_evil:

**Pull request merged**
